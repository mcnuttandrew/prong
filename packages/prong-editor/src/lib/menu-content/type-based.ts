import { SyntaxNode } from "@lezer/common";
import {
  MenuElement,
  nodeToId,
  MenuRow,
  retargetToAppropriateNode,
} from "../compute-menu-contents";

import { boundCheck } from "../modify-json";
import { simpleParse } from "../utils";
import { JSONSchema7 } from "json-schema";

interface TypeComponentProps {
  node: SyntaxNode;
  fullCode: string;
}
type TypeComponent = (props: TypeComponentProps) => MenuRow[];

const directionalMoves = (syntaxNode: SyntaxNode): MenuElement[] => {
  const outputDirections: MenuElement[] = [];
  let node = syntaxNode;
  if (syntaxNode.type.name === "PropertyName") {
    node = syntaxNode.parent!;
  }

  const bounds = boundCheck(node);
  if (!bounds.isFirst) {
    outputDirections.push({
      type: "button",
      content: "Swap with prev",
      onSelect: { type: "decreaseItemIdx", nodeId: nodeToId(node) },
    });
    // {
    //   type: "button",
    //   content: "Set item as Last",
    //   onSelect: { type: "moveItemToEnd" },
    // },
  }
  if (!bounds.isLast) {
    outputDirections.push({
      type: "button",
      content: "Swap with next",
      onSelect: { type: "increaseItemIdx", nodeId: nodeToId(node) },
    });
    // {
    //   type: "button",
    //   content: "Set item as First",
    //   onSelect: { type: "moveItemToStart" },
    // },
  }
  return outputDirections;
};

const PropertyNameComponent: TypeComponent = (props) => {
  const { node } = props;
  return [
    {
      label: "Utils",
      elements: [
        {
          type: "button",
          content: "remove key",
          onSelect: { type: "removeObjectKey", nodeId: nodeToId(node) },
        },
        ...directionalMoves(node),
      ],
    },
  ];
};

const PropertyValueComponent: TypeComponent = (props) => {
  const { node } = props;
  return [
    {
      label: "Utils",
      elements: [
        // { type: "display", content: fullCode.slice(node.from, node.to) },
        {
          type: "button",
          content: "remove key",
          onSelect: {
            type: "removeObjectKey",
            nodeId: nodeToId(node.parent?.firstChild!),
          },
        },
        ...directionalMoves(node.parent?.firstChild!),
      ],
    },
  ];
};

const ObjectComponent: TypeComponent = ({ node }) => [
  {
    label: "Add Field",
    elements: [
      {
        type: "free-input",
        label: "Add field",
        onSelect: {
          type: "addObjectKey",
          nodeId: nodeToId(node),
          payload: { key: "$$INPUT_BIND", value: "null" },
        },
      },
    ],
  },
];

function createObjectMatchingInput(
  code: string,
  node: SyntaxNode
): string | false {
  const x = simpleParse(code.slice(node.from, node.to), undefined);
  if (!Array.isArray(x)) {
    return false;
  }
  const childTypes = new Set(x.map((el) => typeof el));
  if (childTypes.size > 1 || !childTypes.has("object")) {
    return false;
  }
  const keys = Array.from(new Set(x.flatMap((el) => Object.keys(el))));
  return JSON.stringify(Object.fromEntries(keys.map((el) => [el, ""])));
}

const retargetForArray = (node: SyntaxNode): SyntaxNode => {
  if (node.type.name === "[" || node.type.name === "]") {
    return node.parent?.lastChild?.prevSibling!;
  }
  if (node.type.name === "Array") {
    return node.lastChild?.prevSibling!;
  }
  return node;
};

const ArrayComponent: TypeComponent = (props) => {
  const elements = [
    { label: "boolean", value: "false" },
    { label: "number", value: "0" },
    { label: "string", value: '""' },
    { label: "object", value: "{ } " },
    { label: "array", value: "[ ] " },
  ];
  const targetNode = retargetForArray(props.node);
  const simpleObject = createObjectMatchingInput(props.fullCode, targetNode);
  if (simpleObject) {
    elements.push({ label: "inferred object", value: simpleObject });
  }
  return [
    {
      label: "Insert",
      elements: elements.map(({ label, value }) => ({
        type: "button",
        content: label,
        onSelect: {
          type: "addElementAsSiblingInArray",
          payload: value,
          nodeId: nodeToId(targetNode),
        },
      })),
    },
  ];
};

const bracketComponent: TypeComponent = (props) =>
  ArrayComponent({ ...props, node: props.node.parent! });

const curlyBracketComponent: TypeComponent = (props) =>
  ObjectComponent({ ...props, node: props.node.parent! });

const makeSimpleComponent: (x: string) => TypeComponent =
  (_content) => (_props) => {
    return [
      // {
      //   label: "Inferred JSON Type",
      //   elements: [{ type: "display", content }],
      // },
    ];
  };
const ParentIsPropertyComponent = makeSimpleComponent("hi property");

const ParentIsArrayComponent: TypeComponent = ({ node }) => {
  return [
    {
      label: "Utils",
      elements: [
        {
          type: "button",
          content: "Remove Item",

          onSelect: { type: "removeElementFromArray", nodeId: nodeToId(node) },
        },
        ...directionalMoves(node),
      ],
    },
  ];
};

const BooleanComponent: TypeComponent = ({ node }) => [
  {
    label: "Switch to",
    elements: ["true", "false"].map((payload) => ({
      type: "button",
      content: payload,
      onSelect: { type: "simpleSwap", nodeId: nodeToId(node), payload },
    })),
  },
];

const simpleLiteral: TypeComponent = () => {
  return [];
  // disabled bc it adds too much noise and not enough content
  // const elements: MenuElement[] = Object.entries(literalTypes)
  //   .filter(([key]) => key.toLowerCase() !== node.type.name.toLowerCase())
  //   .map(([key, value]) => {
  //     return {
  //       type: "button",
  //       content: key,
  //       onSelect: {
  //         type: "simpleSwap",
  //         nodeId: nodeToId(node),
  //         payload: value,
  //       },
  //     };
  //   });
  // return [{ label: "Switch to", elements }];
};

const composeComponents =
  (a: TypeComponent, b: TypeComponent) => (props: TypeComponentProps) =>
    [...a(props), ...b(props)];

const typeBasedComponents: Record<string, TypeComponent> = {
  Object: ObjectComponent,
  PropertyName: PropertyNameComponent,
  PropertyValue: PropertyValueComponent,
  Array: ArrayComponent,
  String: simpleLiteral,
  Number: simpleLiteral,
  Null: simpleLiteral,
  False: composeComponents(simpleLiteral, BooleanComponent),
  True: composeComponents(simpleLiteral, BooleanComponent),

  ...Object.fromEntries(["[", "]"].map((el) => [el, bracketComponent])),
  ...Object.fromEntries(["{", "}"].map((el) => [el, curlyBracketComponent])),
  "⚠": makeSimpleComponent("⚠"),
};

export function evalTypeBasedContent(
  syntaxNode: SyntaxNode,
  schemaChunk: JSONSchema7[],
  code: string
): MenuRow[] {
  const componentProps = { node: syntaxNode, fullCode: code };
  const type = syntaxNode.type.name;
  const output: MenuRow[] = [];

  const typeBasedProperty: TypeComponent | null = typeBasedComponents[type];

  if (typeBasedProperty) {
    typeBasedProperty(componentProps).forEach((x) => output.push(x));
  } else if (!schemaChunk) {
    console.log("missing imp for", type);
  } else if (!typeBasedProperty) {
    console.log("missing type imp for", type);
  }
  // handle the weird case of property vaues
  if (
    syntaxNode.parent?.type.name === "Property" &&
    syntaxNode.type.name !== "PropertyName"
  ) {
    PropertyValueComponent(componentProps).forEach((x) => output.push(x));
  }

  return output || [];
}

const parentResponses: Record<string, TypeComponent> = {
  Property: ParentIsPropertyComponent,
  Array: ParentIsArrayComponent,
};

export function evalParentBasedContent(
  syntaxNode: SyntaxNode,
  schemaChunk: JSONSchema7[],
  code: string
) {
  const parentType = retargetToAppropriateNode(syntaxNode).parent!.type.name;
  const parentProp = parentResponses[parentType];
  const output = (schemaChunk || []).flatMap((chunk) => {
    const componentProps = {
      content: chunk,
      node: syntaxNode,
      fullCode: code,
    };
    return parentProp ? parentProp(componentProps) : [];
  });
  return output || [];
}
