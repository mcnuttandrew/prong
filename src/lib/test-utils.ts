import { SyntaxNode } from "@lezer/common";
import { parser } from "@lezer/json";
// could also write one that is find by text content?
export function findNodeByLocation(
  text: string,
  from: number,
  to: number
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      if (node.from === from && node.to === to) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}

export function findNodeByText(
  text: string,
  matchText: string
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      const testText = text.slice(node.from, node.to);
      if (testText === matchText) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}
