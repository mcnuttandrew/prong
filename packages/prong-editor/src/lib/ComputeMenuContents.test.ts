import { generateMenuContent } from "./compute-menu-contents";
import { findNodeByText } from "./test-utils";
import { createNodeMap } from "./utils";
import { vegaCode } from "../../../../sites/docs/src/examples/example-data";
import { materializeAnyOfOption } from "./menu-content/schema-based";
// eslint-disable-next-line @typescript-eslint/no-var-requires
const VegaSchema = require("../../../../sites/docs/src/constants/vega-schema.json");
const schema = {
  $id: "https://example.com/arrays.schema.json",
  $schema: "https://json-schema.org/draft/2020-12/schema",
  description: "A representation of a person, company, organization, or place",
  type: "object",
  properties: {
    fruits: {
      type: "array",
      items: {
        type: "string",
      },
    },
    vegetables: {
      type: "array",
      items: { $ref: "#/$defs/veggie" },
    },
  },
  $defs: {
    veggie: {
      type: "object",
      required: ["veggieName", "veggieLike"],
      properties: {
        veggieName: {
          type: "string",
          description: "The name of the vegetable.",
        },
        veggieLike: {
          type: "boolean",
          description: "Do I like this vegetable?",
        },
      },
    },
  },
};

const exampleData = `{
  "fruits": [ "apple", "orange", "pear" ],
  "vegetables": [
    {
      "veggieName": "potato",
      "veggieLike": true
    },
    {
      "veggieName": "broccoli",
      "veggieLike": false
    }
  ]
}`;

test("generateMenuContent - fruit", async () => {
  const veg = findNodeByText(exampleData, `"vegetables"`)!;
  const fAndVNodeMap = await createNodeMap(schema, exampleData);
  expect(fAndVNodeMap).toMatchSnapshot();
  const menuContent = generateMenuContent(veg, fAndVNodeMap, exampleData);
  expect(menuContent).toMatchSnapshot();
});

test("generateMenuContent - vega", async () => {
  const target = findNodeByText(vegaCode, `"transform"`)!.parent?.lastChild
    ?.firstChild?.nextSibling!;
  const nodeMap = await createNodeMap(VegaSchema, vegaCode);
  // expect(nodeMap).toMatchSnapshot();
  const menuContent = generateMenuContent(target, nodeMap, vegaCode);
  expect(menuContent).toMatchSnapshot();
});

test("materializeAnyOfOption", () => {
  const literalSchema = { type: "string", $$labeledType: "role" };
  expect(materializeAnyOfOption(literalSchema as any)).toBe('""');

  const simpleOneOf = {
    $$labeledType: "style",
    $$refName: "#/definitions/style",
    oneOf: [{ type: "string" }, { type: "array", items: { type: "string" } }],
  };
  expect(materializeAnyOfOption(simpleOneOf as any)).toBe('""');

  const simpleAnyOf = {
    $$labeledType: "style",
    $$refName: "#/definitions/style",
    anyOf: [{ type: "string" }, { type: "array", items: { type: "string" } }],
  };
  expect(materializeAnyOfOption(simpleAnyOf as any)).toBe('""');

  const clip = {
    $$labeledType: "clip",
    $$refName: "#/definitions/markclip",
    oneOf: [
      {
        $$refName: "#/definitions/booleanOrSignal",
        oneOf: [
          { type: "boolean" },
          {
            $$refName: "#/definitions/signalRef",
            type: "object",
            properties: {
              signal: { type: "string", $$labeledType: "signal" },
            },
            required: ["signal"],
            $$labeledType: "signalRef",
          },
        ],
        $$labeledType: "booleanOrSignal",
      },
      {
        type: "object",
        properties: {
          path: {
            $$labeledType: "path",
            $$refName: "#/definitions/stringOrSignal",
            oneOf: [
              { type: "string" },
              {
                $$refName: "#/definitions/signalRef",
                type: "object",
                properties: {
                  signal: { type: "string", $$labeledType: "signal" },
                },
                required: ["signal"],
                $$labeledType: "signalRef",
              },
            ],
          },
        },
        required: ["path"],
        additionalProperties: false,
      },
      {
        type: "object",
        properties: {
          sphere: {
            $$labeledType: "sphere",
            $$refName: "#/definitions/stringOrSignal",
            oneOf: [
              { type: "string" },
              {
                $$refName: "#/definitions/signalRef",
                type: "object",
                properties: {
                  signal: { type: "string", $$labeledType: "signal" },
                },
                required: ["signal"],
                $$labeledType: "signalRef",
              },
            ],
          },
        },
        required: ["sphere"],
        additionalProperties: false,
      },
    ],
  };
  expect(materializeAnyOfOption(clip as any)).toBe("true");
});
