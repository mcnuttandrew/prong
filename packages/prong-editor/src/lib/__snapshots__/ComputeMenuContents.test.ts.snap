// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMenuContent - fruit 1`] = `
{
  "0-206": [
    {
      "$defs": {
        "veggie": {
          "properties": {
            "veggieLike": {
              "$$labeledType": "veggieLike",
              "description": "Do I like this vegetable?",
              "type": "boolean",
            },
            "veggieName": {
              "$$labeledType": "veggieName",
              "description": "The name of the vegetable.",
              "type": "string",
            },
          },
          "required": [
            "veggieName",
            "veggieLike",
          ],
          "type": "object",
        },
      },
      "$id": "https://example.com/arrays.schema.json",
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "description": "A representation of a person, company, organization, or place",
      "properties": {
        "fruits": {
          "$$labeledType": "fruits",
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "vegetables": {
          "$$labeledType": "vegetables",
          "items": {
            "$$refName": "#/$defs/veggie",
            "properties": {
              "veggieLike": {
                "$$labeledType": "veggieLike",
                "description": "Do I like this vegetable?",
                "type": "boolean",
              },
              "veggieName": {
                "$$labeledType": "veggieName",
                "description": "The name of the vegetable.",
                "type": "string",
              },
            },
            "required": [
              "veggieName",
              "veggieLike",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
  ],
  "119-123": [
    {
      "$$labeledType": "veggieLike",
      "description": "Do I like this vegetable?",
      "type": "boolean",
    },
  ],
  "135-200": [
    {
      "$$refName": "#/$defs/veggie",
      "properties": {
        "veggieLike": {
          "$$labeledType": "veggieLike",
          "description": "Do I like this vegetable?",
          "type": "boolean",
        },
        "veggieName": {
          "$$labeledType": "veggieName",
          "description": "The name of the vegetable.",
          "type": "string",
        },
      },
      "required": [
        "veggieName",
        "veggieLike",
      ],
      "type": "object",
    },
  ],
  "14-43": [
    {
      "$$labeledType": "fruits",
      "items": {
        "type": "string",
      },
      "type": "array",
    },
  ],
  "157-167": [
    {
      "$$labeledType": "veggieName",
      "description": "The name of the vegetable.",
      "type": "string",
    },
  ],
  "16-23": [
    {
      "type": "string",
    },
  ],
  "189-194": [
    {
      "$$labeledType": "veggieLike",
      "description": "Do I like this vegetable?",
      "type": "boolean",
    },
  ],
  "25-33": [
    {
      "type": "string",
    },
  ],
  "35-41": [
    {
      "type": "string",
    },
  ],
  "61-204": [
    {
      "$$labeledType": "vegetables",
      "items": {
        "$$refName": "#/$defs/veggie",
        "properties": {
          "veggieLike": {
            "$$labeledType": "veggieLike",
            "description": "Do I like this vegetable?",
            "type": "boolean",
          },
          "veggieName": {
            "$$labeledType": "veggieName",
            "description": "The name of the vegetable.",
            "type": "string",
          },
        },
        "required": [
          "veggieName",
          "veggieLike",
        ],
        "type": "object",
      },
      "type": "array",
    },
  ],
  "67-129": [
    {
      "$$refName": "#/$defs/veggie",
      "properties": {
        "veggieLike": {
          "$$labeledType": "veggieLike",
          "description": "Do I like this vegetable?",
          "type": "boolean",
        },
        "veggieName": {
          "$$labeledType": "veggieName",
          "description": "The name of the vegetable.",
          "type": "string",
        },
      },
      "required": [
        "veggieName",
        "veggieLike",
      ],
      "type": "object",
    },
  ],
  "89-97": [
    {
      "$$labeledType": "veggieName",
      "description": "The name of the vegetable.",
      "type": "string",
    },
  ],
}
`;

exports[`generateMenuContent - fruit 2`] = `
[
  {
    "elements": [
      {
        "content": "{"veggieName":"","veggieLike":true}",
        "onSelect": {
          "nodeId": "135-200",
          "payload": "{"veggieName":"","veggieLike":true}",
          "type": "addElementAsSiblingInArray",
        },
        "type": "button",
      },
    ],
    "label": "Insert",
  },
  {
    "elements": [
      {
        "content": "Empty Array",
        "onSelect": {
          "nodeId": "47-59",
          "payload": "[]",
          "type": "simpleSwap",
        },
        "type": "button",
      },
    ],
    "label": "Switch to",
  },
  {
    "elements": [
      {
        "content": "remove key",
        "onSelect": {
          "nodeId": "47-59",
          "type": "removeObjectKey",
        },
        "type": "button",
      },
      {
        "content": "Swap with prev",
        "onSelect": {
          "nodeId": "47-204",
          "type": "decreaseItemIdx",
        },
        "type": "button",
      },
    ],
    "label": "Utils",
  },
]
`;

exports[`generateMenuContent - vega 1`] = `
[
  {
    "elements": [
      {
        "content": "as",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""as"",
            "value": "{}",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "base",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""base"",
            "value": "0",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "divide",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""divide"",
            "value": "{}",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "interval",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""interval"",
            "value": "true",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "maxbins",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""maxbins"",
            "value": "0",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "minstep",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""minstep"",
            "value": "0",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "name",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""name"",
            "value": """",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "span",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""span"",
            "value": "0",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "content": "steps",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": ""steps"",
            "value": "{}",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      {
        "label": "Add field",
        "onSelect": {
          "nodeId": "721-938",
          "payload": {
            "key": "$$INPUT_BIND",
            "value": "null",
          },
          "type": "addObjectKey",
        },
        "type": "free-input",
      },
    ],
    "label": "Add Field",
  },
  {
    "elements": [
      {
        "content": "{"type":"aggregate"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"aggregate"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"bin","field":"{ } ","extent":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"bin","field":"{ } ","extent":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"collect"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"collect"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"contour","size":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"contour","size":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"countpattern","field":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"countpattern","field":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"cross"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"cross"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"crossfilter","fields":"[ ] ","query":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"crossfilter","fields":"[ ] ","query":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"density"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"density"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"dotbin","field":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"dotbin","field":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"extent","field":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"extent","field":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"filter","expr":""}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"filter","expr":""}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"flatten","fields":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"flatten","fields":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"fold","fields":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"fold","fields":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"force"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"force"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"formula","expr":"","as":""}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"formula","expr":"","as":""}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"geojson"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"geojson"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"geopath"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"geopath"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"geopoint","projection":"","fields":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"geopoint","projection":"","fields":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"geoshape"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"geoshape"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"graticule"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"graticule"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"heatmap"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"heatmap"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"identifier","as":""}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"identifier","as":""}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"impute","field":"{ } ","key":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"impute","field":"{ } ","key":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"isocontour"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"isocontour"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"joinaggregate"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"joinaggregate"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"kde","field":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"kde","field":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"kde2d","size":"[ ] ","x":"{ } ","y":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"kde2d","size":"[ ] ","x":"{ } ","y":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"label","size":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"label","size":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"linkpath"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"linkpath"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"loess","x":"{ } ","y":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"loess","x":"{ } ","y":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"lookup","from":"","key":"{ } ","fields":"[ ] "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"lookup","from":"","key":"{ } ","fields":"[ ] "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"nest"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"nest"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"pack"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"pack"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"partition"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"partition"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"pie"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"pie"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"pivot","field":"{ } ","value":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"pivot","field":"{ } ","value":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"project"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"project"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"quantile","field":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"quantile","field":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"regression","x":"{ } ","y":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"regression","x":"{ } ","y":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"resolvefilter","ignore":"0","filter":"null"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"resolvefilter","ignore":"0","filter":"null"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"sample"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"sample"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"sequence","start":"0","stop":"0"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"sequence","start":"0","stop":"0"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"stack"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"stack"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"stratify","key":"{ } ","parentKey":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"stratify","key":"{ } ","parentKey":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"timeunit","field":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"timeunit","field":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"tree"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"tree"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"treelinks"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"treelinks"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"treemap"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"treemap"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"voronoi","x":"{ } ","y":"{ } "}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"voronoi","x":"{ } ","y":"{ } "}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"window"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"window"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
      {
        "content": "{"type":"wordcloud"}",
        "onSelect": {
          "nodeId": "721-938",
          "payload": "{"type":"wordcloud"}",
          "type": "simpleSwap",
        },
        "type": "button",
      },
    ],
    "label": "Switch to",
  },
  {
    "elements": [
      {
        "content": "Remove Item",
        "onSelect": {
          "nodeId": "721-938",
          "type": "removeElementFromArray",
        },
        "type": "button",
      },
      {
        "content": "Swap with next",
        "onSelect": {
          "nodeId": "721-938",
          "type": "increaseItemIdx",
        },
        "type": "button",
      },
    ],
    "label": "Utils",
  },
]
`;
